src/promtail.py

```diff
@@ -0,0 +1,208 @@
+        zip_sha256: The SHA256 hash of the Promtail zip file.
+        bin_sha256: The SHA256 hash of the Promtail binary.
+    """
+
+    url: str
+    zip_sha256: str
+    bin_sha256: str
+
+
+@dataclass
+class Config:
+    """Configuration options for Promtail.
+
+    Attrs:
+        loki_endpoint: The Loki endpoint to send logs to.
+        proxies: Proxy settings.
+        promtail_download_info: Information about the Promtail download.
+    """
+
+    loki_endpoint: str
+    proxies: Optional[ProxyConfig]
+    promtail_download_info: PromtailDownloadInfo
+
+
+def _is_installed(promtail_download_info: PromtailDownloadInfo) -> bool:
+    """Check if Promtail is installed."""
+    return (
+        # pflake8 complains about the line break before the binary operator
+        # but this is how black formats it.
+        PROMTAIL_BINARY_PATH.exists()
+        and PROMTAIL_BINARY_SHA_PATH.exists()  # noqa: W503
+        and PROMTAIL_BINARY_SHA_PATH.read_text(encoding="utf-8")  # noqa: W503
+        == promtail_download_info.bin_sha256  # noqa: W503
+    )
+
+
+def _sha256sums_matches(file_path: Path, sha256sum: str) -> bool:
+    """Check whether a file's sha256sum matches or not with a specific sha256sum.
+
+    Args:
+        file_path: The path to the file to check.
+        sha256sum: The sha256sum against which we want to verify.
+
+    Returns:
+        a boolean representing whether a file's sha256sum matches or not with
+        a specific sha256sum.
+    """
+    try:
+        file_bytes = file_path.read_bytes()
+        result = sha256(file_bytes).hexdigest()
+
+        if result != sha256sum:
+            logger.debug(
+                "File sha256sum mismatch, expected:'%s{sha256sum}' but got '%s'",
+                sha256sum,
+                result,
+            )
+            return False
+        return True
+    except FileNotFoundError:
+        logger.error("File: '%s' could not be opened", file_path)
+        return False
+
+
+def _install(promtail_download_info: PromtailDownloadInfo) -> None:
+    if not promtail_download_info.url.startswith(PROMTAIL_BASE_URL):
+        raise PromtailInstallationError(
+            f"Unknown Promtail download URL {promtail_download_info.url}. "
+            f"Must start with {PROMTAIL_BASE_URL}"
+        )
+    response = requests.get(promtail_download_info.url, timeout=300)
+    response.raise_for_status()
+
+    with open(PROMTAIL_BINARY_PATH.with_suffix(".gz"), "wb") as file:
+        logger.info("Writing Promtail binary zip to %s", PROMTAIL_BINARY_PATH.with_suffix(".gz"))
+        file.write(response.content)
+    if not _sha256sums_matches(
+        PROMTAIL_BINARY_PATH.with_suffix(".gz"), promtail_download_info.zip_sha256
+    ):
+        raise PromtailInstallationError(
+            f"Promtail zip file sha256sum mismatch, expected: {promtail_download_info.zip_sha256}"
+        )
+    with gzip.open(PROMTAIL_BINARY_PATH.with_suffix(".gz"), "rb") as file:
```

Comment:
`PROMTAIL_BINARY_PATH.with_suffix(".gz")` is repeated multiple times, worthwhile storing it in a variable